ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/LMDB.hs

```diff
@@ -0,0 +1,545 @@
+  deriving (Show, Eq)
+
+data TraceTableOp = TTO
+  deriving (Show, Eq)
+
+data TDBTableOp
+  = TTORead
+  | TTOInit
+  | TTOWrite
+  deriving (Show)
+
+{-------------------------------------------------------------------------------
+ Errors
+-------------------------------------------------------------------------------}
+
+data DbErr = DbErrStr !String
+  | DbErrNoSettings
+  | DbErrNonMonotonicSeq !(WithOrigin SlotNo) !(WithOrigin SlotNo)
+  | DbErrNoDbNamed !String
+  | DbErrBadDynamic !String
+  | DbErrInitialisingNonEmpty !String
+  | DbErrBadRead
+  | DbErrBadRangeRead
+  | DbErrDbExists !FS.FsPath
+  | DbErrDbDoesntExist !FS.FsPath
+  deriving stock (Show)
+
+instance Exn.Exception DbErr
+
+{-------------------------------------------------------------------------------
+  Database definition
+-------------------------------------------------------------------------------}
+
+data Db l = Db
+  { dbEnv           :: !(LMDB.Environment LMDB.ReadWrite) -- ^ The LMDB environment is a pointer to the directory that contains the DB.
+  , dbSettings      :: !(LMDB.Database () DbState) -- ^ A database with only one key and one value, for the current sequence number on the DB.
+  , dbBackingTables :: !(LedgerTables l LMDBMK) -- ^ The LMDB database with the key-value store
+  , dbFilePath      :: !FilePath
+  , dbTracer        :: !(Trace.Tracer IO TraceDb)
+  , dbOpenHandles   :: !(IOLike.TVar IO (Map Int ValueHandle))
+  }
+
+type LMDBLimits = LMDB.Limits
+
+defaultLMDBLimits :: LMDBLimits
+defaultLMDBLimits = LMDB.defaultLimits
+  { LMDB.mapSize = 6_000_000_000
+  , LMDB.maxDatabases = 10 -- We use 1 database per field + one for the state (i.e. sidetable)
+  , LMDB.maxReaders = 16
+  }
+
+newtype DbState  = DbState
+  { dbsSeq :: WithOrigin SlotNo  -- TODO a version field
+  } deriving (Show, Generic)
+
+instance S.Serialise DbState
+
+newtype LmdbBox a = LmdbBox a
+  deriving newtype (Eq, Show, Ord)
+
+instance (CBOR.ToCBOR a, CBOR.FromCBOR a) => S.Serialise (LmdbBox a) where
+  encode (LmdbBox a) = CBOR.toCBOR a
+  decode = LmdbBox <$> CBOR.fromCBOR
+
+coerceDatabase :: LMDB.Database k v -> LMDB.Database (LmdbBox k) (LmdbBox v)
+coerceDatabase = unsafeCoerce
+
+-- | ValueHandles hold an Async which is holding a transaction open.
+data ValueHandle = ValueHandle
+  { vhClose :: !(IO ())
+  , vhSubmit :: !(forall a. LMDB.Transaction LMDB.ReadOnly a -> IO (Maybe a))
+  -- , vhRefCount :: !(IOLike.TVar IO Int)
+  }
+
+{-------------------------------------------------------------------------------
+ LMDB Interface
+-------------------------------------------------------------------------------}
+
+lmdbReadTable :: LedgerConstraint k v
+  => LMDB.Database k v
+  -> HD.UtxoKeys k v
+  -> LMDB.Transaction mode (HD.UtxoValues k v)
+lmdbReadTable db0 (HD.UtxoKeys keys) =
+    HD.UtxoValues <$> foldlM' go Map.empty (Set.toList keys)
+  where
+    db = coerceDatabase db0
+    go m k = LMDB.get db (LmdbBox k) <&> \case
```

Comment:
If the following makes sense we should probably add a TODO and JIRA ticket: I wonder if we should be issuing a single C call to get a bunch of keys in one go.